### PR TITLE
File Based Configuration

### PR DESCRIPTION
**Pull request by opalmer**
This PR converts the calls to the `read_env_*` to configuration lookups and:

* Removes `pyfarm.models.core.cfg` as the central location for configuring table 
* Removes many of the less commonly used environment variables.  We can any back in that are in use today of course.  Any that we wish to keep or add in the future are also supported at load time in `pyfarm.master.config`.
* Adds documentation to some of the classes which don't have them as well as the values in the config.
* Cleans up some formatting of some of the table classes.  I thought of splitting this off into another CL but I would had to reformat several of the classes anyway to fit within the line length limit.  For consistency some of the other table classes were updated too.

### CODE REVIEW

**Comment by guidow** [comment]
I haven't yet looked through all of the changes, but I like the idea, and I like what I've seen so far. Configuring pyfarm-master purely through environment variables has always been a bit on the weird side...

In any case, this PR will need some very thorough review and testing, since it has the potential to break a lot of things if something is overlooked somewhere.

**Comment by opalmer** [comment]
> I haven't yet looked through all of the changes, but I like the idea, and I like what I've seen so far. Configuring pyfarm-master purely through environment variables has always been a bit on the weird side...

Awesome thanks for the feedback, glad to hear this is going in the right direction.  

> In any case, this PR will need some very thorough review and testing, since it has the potential to break a lot of things if something is overlooked somewhere.

Couldn't agree more.  I'm reviewing this quite a bit myself as I make changes but it would still be fairly easy to make a subtle change without breaking things.  That's one reason I included some of the main environment variables too, it's making it easier to transition and it should replicate the defaults more closely in the short term.  There were also a couple of places I've been thinking about adding or updating tests too to make sure we're covering some of the changes in this PR.

**Comment by opalmer** [comment]
@guidow, I think this is ready for review and testing.  I'm not finding any more calls to read_env or any of it's sibling functions.

File: pyfarm/master/etc/master.yml [hunk 1]
```diff
@@ -12,7 +12,206 @@
...
+# The directory to store downloaded logs in.
+#
+# **Production Note**: For production it's probably best if these are kept
+# in a persistent location rather than $temp.
+tasklogs_dir: ${temp}/task_logs
+
+
+# The address the Flask application should listen on.  By default
+# this will only listen locally but could be changed to listen on
+# a specific adatper or `0.0.0.0` for all addresses.
```
Comment (guidow) [inline pyfarm/master/etc/master.yml]: This should point out that this setting is only for flask's standalone test webserver

File: pyfarm/master/etc/master.yml [hunk 2]
```diff
@@ -12,7 +12,206 @@
...
+dev_db_create_all: false
+
+
+# When true the application will be instanced as 'app' in the
+# pyfarm.master.entrypoints module.  When running behind something
+# like uwsgi this should be true.
+instance_application: false
+
+##
+## BEGIN Queue defautls
```
Comment (guidow) [inline pyfarm/master/etc/master.yml]: Typo

File: pyfarm/scheduler/etc/scheduler.yml [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+
+# Used when polling agents to determine if an agent is considered
+# offline or not after a given period of time without communication.  The keys
+# and values here are passed into a `timedelta` object as keywords.
+poll_offline_agents_interval:
+  hours: 2
+
+
+# A directory where lock files for the scheuler can be found.
```
Comment (guidow) [inline pyfarm/scheduler/etc/scheduler.yml]: Typo

**Comment by guidow** [comment]
When starting the master locally with those changes, I get this error:

    2015-05-08 18:11:29 DEBUG    - pf.core.config  - Created '/tmp/pyfarm/master'
    2015-05-08 18:11:29 WARNING  - pf.core.config  - No environment was provided to be populated by the configuration file(s)
    2015-05-08 18:11:29 INFO     - pf.core.config  - Loaded configuration file(s): ['/home/guido/git/pyfarm/pyfarm-master/pyfarm/master/etc/master.yml']
    2015-05-08 18:11:29 DEBUG    - pf.core.config  - Created '/tmp/pyfarm/pyfarm.models'
    2015-05-08 18:11:29 WARNING  - pf.core.config  - No environment was provided to be populated by the configuration file(s)
    2015-05-08 18:11:29 INFO     - pf.core.config  - Loaded configuration file(s): ['/home/guido/git/pyfarm/pyfarm-master/pyfarm/models/etc/models.yml']
    2015-05-08 18:11:29 DEBUG    - pf.core.config  - Created '/tmp/pyfarm/pyfarm.scheduler'
    2015-05-08 18:11:30 WARNING  - pf.core.config  - No environment was provided to be populated by the configuration file(s)
    2015-05-08 18:11:30 INFO     - pf.core.config  - Loaded configuration file(s): ['/home/guido/git/pyfarm/pyfarm-master/pyfarm/scheduler/etc/scheduler.yml']
    2015-05-08 18:11:30 INFO     - pf.core.config  - read_env('PYFARM_DEV_LISTEN_ON_WILDCARD'): 'True'
    2015-05-08 18:11:30 INFO     - pf.core.config  - read_env('PYFARM_BASE_URL'): 'http://127.0.0.1:5000'
    2015-05-08 18:11:30 INFO     - pf.core.config  - read_env('PYFARM_DEV_ALLOW_AGENT_LOOPBACK_ADDRESSES'): 'True'
    2015-05-08 18:11:30 INFO     - pf.core.config  - read_env('PYFARM_FARM_NAME'): 'Dev'
    Traceback (most recent call last):
      File "/var/tmp/virtualenv/pyfarm-master-python3/bin/pyfarm-master", line 9, in <module>
        load_entry_point('pyfarm.master==0.8.6', 'console_scripts', 'pyfarm-master')()
      File "/home/guido/git/pyfarm/pyfarm-master/pyfarm/master/entrypoints.py", line 654, in run_master
        app.run(host=config.get("flask_listen_address"))
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/Flask-0.10.1-py3.3.egg/flask/app.py", line 772, in run
        run_simple(host, port, self, **options)
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/Werkzeug-0.10.4-py3.3.egg/werkzeug/serving.py", line 607, in run_simple
        if ':' in display_hostname:
    TypeError: argument of type 'bool' is not iterable

Generally speaking, there is way too much stuff in this PR that is not strictly related to, let alone necessary for switching to a file based configuration, like the sweeping code formatting changes in the database models, or reworking how the scheduler sends emails. (Not that that's a bad idea per se.) This makes review unnecessarily hard and time consuming and merging unnecessarily risky.

Please don't merge this yet. I have looked over the code and didn't find anything too objectionable, but it's a lot to keep track of, and I may easily have overlooked something subtle. I would really prefer to have this running in our test environment for a couple of days before this gets merged.

Also, if it's not asking too much, please move the changes that are not directly related to file based configuration to other branches. This pull request is big enough already even without those.

Also, at this point in time, a rebase to master might be a good idea.

**Comment by guidow** [comment]
I've tried running the tests locally, and most of the tables in my local sqlite database got dropped.

On the plus side, I get to test restore from backup now...

File: pyfarm/master/config.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        "database": ("PYFARM_DATABASE_URI", read_env_no_log),
+        "timestamp_format": ("PYFARM_TIMESTAMP_FORMAT", read_env),
+        "allow_agents_from_loopback": (
+            "PYFARM_DEV_ALLOW_AGENT_LOOPBACK_ADDRESSES", read_env_bool),
+        "agent_updates_dir": ("PYFARM_AGENT_UPDATES_DIR", read_env),
+        "agent_updates_webdir": ("PYFARM_AGENT_UPDATES_WEBDIR", read_env),
+        "farm_name": ("PYFARM_FARM_NAME", read_env),
+        "tasklogs_dir": ("PYFARM_LOGFILES_DIR", read_env),
+        "flask_listen_address": (
+            "PYFARM_DEV_LISTEN_ON_WILDCARD", env_bool_false),
```
Comment (guidow) [inline pyfarm/master/config.py]: You cannot use "PYFARM_DEV_LISTEN_ON_WILDCARD" as a default for "flask_listen_address". One is a bool, the other is a string that's supposed to represent an IP address.

**Comment by guidow** [comment]
I see this passes all tests on travis, but when I run the tests locally on my machine, the results look like this:

    FAILED (SKIP=2, errors=329, failures=94)

**Comment by opalmer** [comment]
> Please don't merge this yet. I have looked over the code and didn't find anything too objectionable, but it's a lot to keep track of, and I may easily have overlooked something subtle. I would really prefer to have this running in our test environment for a couple of days before this gets merged.

Yeah I don't plan on merging this until you're fully ok with it from a code perspective and from a testing perspective.  It's simply too large of a change to risk doing that (even if some of the other changes are split out).

> Also, if it's not asking too much, please move the changes that are not directly related to file based configuration to other branches.

I'll be happy too.  Like I said in the original description, before I edited it, I realize this one is quite large so I fully expected to need to pair this down.

> I see this passes all tests on travis, but when I run the tests locally on my machine, the results look like this:

Yeah I'm seeing some failures too, only ten though and they all seem to be related to log files.  Three questions for you if you don't mind, trying to see if we're seeing the same failures here:

* Is it a specific group of tests that are failing?
* If you're testing locally with sqlite, what version of sqlalchemy are you using?
* If you are using sqlite, does it make a difference if you move the DB out of the way (assuming you're not using `:memory:`)?

Based on the tests that I see failing I'm guessing either one of the defaults is wrong or the expectation of where the logs are being put in the tests is wrong since the config has consolidated some of the temp paths in the default.

**Comment by opalmer** [comment]
Ok I've split out most of the major changes into the referenced PRs.  That should pair down this PR quite a bit.

**Comment by opalmer** [comment]
@guidow, all the broken out changes have been merged and I've also updated to the latest master.  You had any issues with testing this so far?

**Comment by guidow** [comment]
The tests are all successful, but instead of using temporary tables with a random prefix for testing, as the tests did before this, they will now reuse the existing tables – and drop them at the end. This is not good, we should get the old behavior back.

Another problem:
When sending an assignment to an agent, i get this error:

    [2015-05-22 19:10:38,414: ERROR/MainProcess] Task pyfarm.scheduler.tasks.send_tasks_to_agent[2770cf3d-c1b9-412c-b812-d6286c53f430] raised unexpected: InvalidURL('Failed to parse: 1:$50000',)
    Traceback (most recent call last):
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/celery/app/trace.py", line 240, in trace_task
        R = retval = fun(*args, **kwargs)
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/celery/app/trace.py", line 438, in __protected_call__
        return self.run(*args, **kwargs)
      File "/home/guido/git/pyfarm/pyfarm-master/pyfarm/scheduler/tasks.py", line 203, in send_tasks_to_agent
        timeout=AGENT_REQUEST_TIMEOUT)
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/requests/api.py", line 109, in post
        return request('post', url, data=data, json=json, **kwargs)
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/requests/api.py", line 50, in request
        response = session.request(method=method, url=url, **kwargs)
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/requests/sessions.py", line 451, in request
        prep = self.prepare_request(req)
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/requests/sessions.py", line 382, in prepare_request
        hooks=merge_hooks(request.hooks, self.hooks),
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/requests/models.py", line 304, in prepare
        self.prepare_url(url, params)
      File "/var/tmp/virtualenv/pyfarm-master-python3/lib/python3.3/site-packages/requests/models.py", line 357, in prepare_url
        raise InvalidURL(*e.args)
    requests.exceptions.InvalidURL: Failed to parse: $127.0.0.1:$50000

Also, this branch needs rebasing to master.

File: pyfarm/master/etc/master.yml [hunk 5]
```diff
@@ -12,7 +12,207 @@
...
+
+
+# The prefix of the URL from which the API will operate on.  This should
+# not generally be changed unless you are operating different versions
+# of the API at the same time from one web server.
+api_prefix: /api/v1
+
+
+# The URL template we use to communicate with the agent.
+agent_api_url_template: http://${host}:${port}/api/v1
```
Comment (guidow) [inline pyfarm/master/etc/master.yml]: Should probably read ```http://{host}:{port}/api/v1``` (without the dollar signs).

**Comment by guidow** [comment]
Other than this, the master is working like this. Once the problems mentioned above are taken care of, I plan to deploy this branch in our test farm and monitor it for while.

File: pyfarm/master/etc/master.yml (hunk 5 above)
Comment (opalmer) [inline pyfarm/master/etc/master.yml]: Fixed.

**Comment by opalmer** [comment]
> The tests are all successful, but instead of using temporary tables with a random prefix for testing, as the tests did before this, they will now reuse the existing tables – and drop them at the end. This is not good, we should get the old behavior back.

Yes agreed, I'll get this fixed.

File: pyfarm/master/etc/master.yml [hunk 6]
```diff
@@ -12,7 +12,207 @@
...
+
+
+# Enables verbose output of loggers associated with the
+# job queue.
+debug_queue: false
+
+
+# The URL to access the database.  For debugging and development
+# a local SQLite database is used.  For production access should be
+# configured to a more standard SQL database.  MySQL or Postgres are
```
Comment (guidow) [inline pyfarm/master/etc/master.yml]: Awkward sentence structure. Maybe reorder?

File: pyfarm/master/etc/master.yml [hunk 7]
```diff
@@ -12,7 +12,207 @@
...
+login_disabled: false
+
+
+# The amount of time the 'remeber me' cookie should persist.  The keys
+# and values here are passed into a `timedelta` object as keywords.
+cookie_duration:
+  hours: 16
+
+
+# When true json output from te APIs will be reformatted to
```
Comment (guidow) [inline pyfarm/master/etc/master.yml]: Typo: ```te``` -> ```the```.

File: pyfarm/master/etc/master.yml [hunk 8]
```diff
@@ -12,7 +12,207 @@
...
+# like uwsgi this should be true.
+instance_application: false
+
+##
+## BEGIN Queue defaults
+##
+
+# When true the queue will prefer to assign work
+# for jobs which are already running.
+queue_prefer_running_jobs: true
```
Comment (guidow) [inline pyfarm/master/etc/master.yml]: This one should probably go into the scheduler config instead.

File: pyfarm/scheduler/etc/scheduler.yml [hunk 9]
```diff
@@ -13,3 +11,211 @@
...
+# The default address from which all emails from the scheduler will
+# originate.
+from_email: pyfarm@localhost
+
+##
+## BEGIN Email Template Settings
+##
+
+# General note about the settings below.  The brackets, {{ }}
+# are used by the tempating system for string substitution.  For example,
```
Comment (guidow) [inline pyfarm/scheduler/etc/scheduler.yml]: s/tempating/templating

File: pyfarm/scheduler/etc/scheduler.yml [hunk 10]
```diff
@@ -13,3 +11,211 @@
...
+## END Scheduler Settings
+##
+
+##
+## BEGIN Email Server Settings
+##
+
+##
+## END Email Server Settings
+##
```
Comment (guidow) [inline pyfarm/scheduler/etc/scheduler.yml]: This block should be moved after the actual email server settings.

File: pyfarm/scheduler/etc/scheduler.yml [hunk 11]
```diff
@@ -13,3 +11,211 @@
...
+poll_offline_agents_interval:
+  hours: 2
+
+
+# A directory where lock files for the scheuler can be found.
+scheduler_lockfile_base: ${temp}/scheduler_lock
+
+
+# The directory where all task logs should be stored.
+task_logs_dir: ${temp}/task_logs
```
Comment (guidow) [inline pyfarm/scheduler/etc/scheduler.yml]: This should be part of master.yml instead.

**Comment by guidow** [comment]
At least under CentOS 7, when installing with pip, for some reason the default config files get install under /usr/etc/pyfarm/{master,models,scheduler}.yml, but the code will search for them under ```/usr/lib/python2.7/site-packages/pyfarm/master/etc/master.yml``` and similar for models.yml and scheduler.yml.

File: pyfarm/master/etc/master.yml [hunk 12]
```diff
@@ -12,7 +12,207 @@
...
+
+# When true the application will automatically create users in
+# PyFarm's database if they do not exist already.  Setting this
+# to false will cause an exception to be raised if the user in
+# question does not exist.
+autocreate_users: true
+
+
+# When autocreating users, use this email address as a template.  For example:
+#   {username}@example.com
```
Comment (guidow) [inline pyfarm/master/etc/master.yml]: The given example will confuse the yaml parser, leading to errors like this:

    Failed to load '/etc/pyfarm/master/master.yml': while scanning for the next token
    found character '@' that cannot start any token

**Comment by guidow** [comment]
I have opened a pull for request for this branch that addresses some of the problems I encountered with the changed setup.py.

File: pyfarm/scheduler/tasks.py [hunk 13]
```diff
@@ -80,56 +78,29 @@
...
+SMTP_SERVER = config.get("smtp_server")
+SMTP_PORT = config.get("smtp_port")
+SMTP_USER, SMTP_PASSWORD = config.get("smtp_login")
+FROM_ADDRESS = config.get("from_email")
+DEFAULT_SUCCESS_SUBJECT = Template(config.get("success_subject"))
+DEFAULT_SUCCESS_BODY = Template(config.get("success_body"))
+DEFAULT_FAIL_SUBJECT = Template(config.get("failed_subject"))
+DEFAULT_FAIL_BODY = Template(config.get("failed_body"))
+DEFAULT_DELETE_SUBJECT = config.get("deleted_subject")
+DEFAULT_DELETE_BODY = config.get("deleted_body")
```
Comment (guidow) [inline pyfarm/scheduler/tasks.py]: These two should be packaged inside a Template object as well.

File: pyfarm/scheduler/tasks.py [hunk 14]
```diff
@@ -814,12 +787,16 @@ def send_job_completion_mail(job_id, successful=True):
 
 @celery_app.task(ignore_results=True)
 def send_job_deletion_mail(job_id, jobtype_name, job_title, to):
-    message_text = ("%s job %s (id %s) has been deleted.\n\n" %
-                    (jobtype_name, job_title, job_id))
-    message_text += "Sincerely,\n\tThe PyFarm render manager"
+    logger.debug("In send_job_deletion_mail(), job_id: %s, jobtype_name: %s, "
+                 "job_title: %s, to: %s", job_id, jobtype_name, job_title, to)
+    message_text = DEFAULT_DELETE_SUBJECT.format(
```
Comment (guidow) [inline pyfarm/scheduler/tasks.py]: Should use ```DEFAULT_DELETE_BODY``` in this place.

File: pyfarm/scheduler/tasks.py [hunk 15]
```diff
@@ -814,12 +787,16 @@ def send_job_completion_mail(job_id, successful=True):
...
+        job_title=job_title,
+        jobtype_name=jobtype_name,
+        job_id=job_id
+    )
 
     message = MIMEText(message_text)
-    message["Subject"] = ("Job %s deleted" % job_title)
+    message["Subject"] = DEFAULT_DELETE_SUBJECT.format(title=job_title)
```
Comment (guidow) [inline pyfarm/scheduler/tasks.py]: Keyword ```title``` should be ```job_title``` instead.

File: pyfarm/master/etc/master.yml (hunk 12 above)
Comment (opalmer) [inline pyfarm/master/etc/master.yml]: ~~Hmm...That's odd considering it's not a problem with the default config~~ 

Nevermind, you meant it's a problem if someone copies/paste the config as a value...that makes much more sense.  Fixing...

**Comment by opalmer** [comment]
> At least under CentOS 7, when installing with pip, for some reason the default config files get install under /usr/etc/pyfarm/{master,models,scheduler}.yml, but the code will search for them under /usr/lib/python2.7/site-packages/pyfarm/master/etc/master.yml and similar for models.yml and scheduler.yml.

This may be fixed with some of the setup.py changes you've merged.

File: pyfarm/master/etc/master.yml (hunk 6 above)
Comment (opalmer) [inline pyfarm/master/etc/master.yml]: Yeah it is a little odd....2388999 should improve on this some.

**Comment by guidow** [comment]
At the moment, the config code will correctly load and parse ```/etc/pyfarm/master/master.yml```, but not ```/etc/pyfarm/scheduler/scheduler.yml``` or ```/etc/pyfarm/models/models.yml```. This appears to be a bug in pyfarm-core, though.

**Comment by opalmer** [comment]
Merged pyfarm/pyfarm-core#40 today and I'll roll out a release.  Hopefully that PR will fix this problem.

**Comment by guidow** [comment]
This his been running in our test environment for a few days now, and after a few early hiccups, it's running well.

**Comment by opalmer** [comment]
That's great news.  I'd support testing for a little longer if you want but I'm happy to merge this tomorrow too.

**Comment by guidow** [comment]
I gave the changes another once-over, ran the tests again, and I think we're about done here.

**Comment by guidow** [comment]
I think we can merge this now.

**Comment by opalmer** [comment]
Sorry for the delay, merging!

**Pull request merged**